Add health check paths to scheduled update endpoint

zaerl opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes https://github.com/Automattic/jetpack/issues/36982

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Add `health_check_paths` to `wpcom/v2/update-schedules` and save it in `jetpack_scheduled_update_health_check_paths`
* Send `health_check_paths` to `/sites/%d/hosting/scheduled-update`
* `health_check_paths` is not required and defaults to `array()` for retrocompatibility

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

## On WPCOM
1. Add D146273-code
2. Add `\l( $request );` to `WPCOM_REST_API_V2_Endpoint_Atomic_Hosting_Scheduled_Update::update_item`
3. Check `/tmp/php-errors` and ensure the data is passed.

## On Jetpack
1. On Jetpack Beta > WordPress.com Features install `add/su-health-check-paths`.
2. Install REST API console or similar.
3. Force a plugin version to an old one.
4. Create a scheduled update one minute in the future, such as:

`/wpcom/v2/update-schedules?plugins[]=rest-api-console/rest-api-console.php&schedule[interval]=daily&schedule[timestamp]=NNNNN&schedule[health_check_paths][]=test1&schedule[health_check_paths][]=test2`

You can obtain 60 seconds in the future in the dev tools by writing `const d = new Date(); Math.floor( d.getTime() / 1000 ) + 60;`

1. Refresh the plugins page and wait a minute.
2. In the meantime, access `_cli` and `option get jetpack_scheduled_update_health_check_paths` and ensure your paths are there.
3. Try to delete all your scheduled updates and ensure the option is removed.

zaerl committed:
Add health paths to scheduled updates

zaerl committed:
changelog

zaerl committed:
Fix: typo

zaerl committed:
Minor refactoring

zaerl committed:
Add unit tests

zaerl committed:
Set paths as required false

zaerl committed:
Add option deletion

obenland commented:
Before we continue here, let's get https://github.com/Automattic/jetpack/pull/36961 and https://github.com/Automattic/jetpack/pull/36835 in first please. The main class and API endpoint are getting quite crowded.

It would also make it easier to keep health check paths in a separate file for example

zaerl committed:
Add limit to 5 paths

ouikhuan commented on projects/packages/scheduled-updates/src/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-update-schedules.php:
```diff
@@ -651,6 +654,28 @@ public function validate_themes_param( $themes ) {
+	 * @param array $paths List of paths to check.
+	 * @return bool|WP_Error
+	 */
+	public function validate_paths_param( $paths ) {
```
I'm wondering if we should do more validation here, for example, we probably want to allow users to create both absolute URLs and relative URLs, if they use absolute URLs, we will need to make sure the domain matches their site. ~~I'm also not sure if we should do a header check at this stage, otherwise user might input something that's 404 in the first place, what do you think? One caveat of doing the header check here is that it might slow down the original action.  Or we should just let it pass through and wait until later?~~

Updated: This was discussed here p1713800108107349/1713798556.992189-slack-C01A60HCGUA will be handled from the frontend.

zaerl replied:
Now the backend accept paths such as:
1. `https://example.com`
2. `https://example.com/`
3. `https://example.com/test`
4. Empty string
5. `/`
6. `/test`
7. Etc.

ouikhuan commented on projects/packages/scheduled-updates/tests/php/class-wpcom-rest-api-v2-endpoint-update-schedules-test.php:
```diff
@@ -940,6 +909,147 @@ public function test_add_log_unauthorized() {
+	 *
+	 * @covers ::create_item
+	 */
+	public function test_remove_item_with_with_paths() {
```
There's a duplicate `with` typo here.

zaerl replied:
Fixed, thanks.

ouikhuan commented on projects/packages/scheduled-updates/tests/php/class-wpcom-rest-api-v2-endpoint-update-schedules-test.php:
```diff
@@ -940,6 +909,147 @@ public function test_add_log_unauthorized() {
+		$result  = rest_do_request( $request );
+		$this->assertSame( 200, $result->get_status() );
+
+		$option_paths = Scheduled_Updates::get_health_check_paths( $schedule_id_1 );
```
Should this be `$schedule_id_2` instead?

zaerl replied:
You're right, fixed.

ouikhuan commented on projects/packages/scheduled-updates/tests/php/class-wpcom-rest-api-v2-endpoint-update-schedules-test.php:
```diff
@@ -940,6 +909,147 @@ public function test_add_log_unauthorized() {
+		$option_paths = Scheduled_Updates::get_health_check_paths( $schedule_id );
+		$this->assertSame(
+			array(
+				'a_b',
+				'=',
+				'',
+				'test',
+				'*',
+				'ünicode',
```
Should we allow path like only `=`, `*` or `' '` to pass the validation?

zaerl replied:
Now, all the strings filtered away by `esc_url_raw` are not accepted.

ouikhuan commented on projects/packages/scheduled-updates/src/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-update-schedules.php:
```diff
@@ -791,17 +820,25 @@ public function get_object_params() {
+					'health_check_paths' => array(
+						'description'       => 'List of paths to check for site health after the update.',
+						'type'              => 'array',
+						'maxItems'          => 5,
```
Should we add the item type check here as well? Something like `'items' => array( 'type' => 'string' )`. I noticed we do have the string check on the `validate_paths_params check` but it doesn't seem to be triggered at all.  During my tests, I did a request like the following:
```
plugins[]=rest-api-console/rest-api-console.php&schedule[interval]=daily&schedule[timestamp]=1713843069&schedule[health_check_paths][]=test1&schedule[health_check_paths][]=test2&schedule[health_check_paths][]=test3&schedule[health_check_paths][]=test4&schedule[health_check_paths][]=test5
```
It didn't get stopped by the validate check, but later on when it tries to trim the path, the value is an array and not string and throws me a 500 error. This is because the data structure I passed at the first place is incorrect. Eventually, I solved it by adding array index to my request:
```
plugins[]=rest-api-console/rest-api-console.php&schedule[interval]=daily&schedule[timestamp]=1713843069&schedule[health_check_paths][0]=test1&schedule[health_check_paths][1]=test2&schedule[health_check_paths][2]=test3&schedule[health_check_paths][3]=test4&schedule[health_check_paths][4]=test5
```

zaerl replied:
Added, thanks.

zaerl committed:
Add unit tests fix

zaerl committed:
Add items type check

zaerl committed:
Fix: variable typo

zaerl committed:
Move health paths code to a separate class

zaerl committed:
Move health checks unit tests to separate class

zaerl committed:
Fix phan static errors

zaerl committed:
Add new unit tests

zaerl committed:
Add esc_url_raw

zaerl committed:
Minor change

TimBroddin reviewed (dismissed):
Tests well and code looks solid. Nice job!

There's on Phan error you might want to fix before merging.

zaerl committed:
Add warning suppress

zaerl commented:
> Tests well and code looks solid. Nice job!
> 
> There's on Phan error you might want to fix before merging.

I've added a Phan suppress because the new statement must stay there. Added on `class-wpcom-rest-api-v2-endpoint-update-schedules-test.php` as well.

zaerl committed:
Remove Phan ignore statements

zaerl committed:
Update Phan baseline

zaerl committed:
changelog

obenland commented on projects/packages/scheduled-updates/src/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-update-schedules.php:
```diff
@@ -266,6 +267,7 @@ public function create_item( $request ) {
 		}
 
 		$id = Scheduled_Updates::generate_schedule_id( $plugins );
+		Scheduled_Updates_Health_Paths::update( $id, $schedule['health_check_paths'] ?? array() );
```
@zaerl Could I ask you for a follow-up to move the update and clear calls into action callbacks?

zaerl replied:
https://github.com/Automattic/dotcom-forge/issues/6700
